main.c

```diff
@@ -128,51 +128,55 @@ static void apply_profile(struct kanshi_state *state,
 	zwlr_output_configuration_v1_add_listener(config, &config_listener, pending);
 
 	size_t i = 0;
+	struct kanshi_head *head;
 	struct kanshi_profile_output *profile_output;
-	wl_list_for_each(profile_output, &profile->outputs, link) {
-		struct kanshi_head *head = matches[i];
-
-		bool enabled = head->enabled;
-		if (profile_output->fields & KANSHI_OUTPUT_ENABLED) {
-			enabled = profile_output->enabled;
-		}
+	while ((head = matches[i]) != NULL) {
+		wl_list_for_each(profile_output, &profile->outputs, link) {
+			if (!match_profile_output(profile_output, head))
```

Comment:
Hmm, I'm not sure I understand why we re-do the matching here. The profile_output <-> head matching has already been done before-hand when selecting the profile to apply, no need to do it again

Reply:
> Looks like this is fixed by [32666e0](https://github.com/emersion/kanshi/commit/32666e0f584068bb389fbf5ffaab4008956cb07f)

I take it back. It is still broken for me. Here is a backtrace which might help explain the issue:
```
Breakpoint 1, apply_profile (state=0x7fffffffe370, profile=0x55555555e570, matches=0x7fffffffdd30)
    at ../main.c:120
120             if (state->current_profile == profile) {
(gdb) list-print profile->outputs
list@0x55555555e580: {prev = 0x55555555e5b0, next = 0x55555555e630}
node@0x55555555e630: {prev = 0x55555555e580, next = 0x55555555e5b0} #1
node@0x55555555e5b0: {prev = 0x55555555e630, next = 0x55555555e580} #2
(gdb) print *container_of(0x55555555e630, struct kanshi_profile_output, link)
$6 = {name = 0x55555555e670 "Dell Inc. DELL UP3216Q 6XF0D659376", fields = 11, link = {
    prev = 0x55555555e580, next = 0x55555555e5b0}, enabled = true, mode = {width = 3840,
    height = 2160, refresh = 0}, position = {x = 0, y = 0}, scale = 2,
  transform = WL_OUTPUT_TRANSFORM_NORMAL}
(gdb) print *container_of(0x55555555e5b0, struct kanshi_profile_output, link)
$7 = {name = 0x55555555e5f0 "Sharp Corporation 0x1484 0x00000000", fields = 1, link = {
    prev = 0x55555555e630, next = 0x55555555e580}, enabled = false, mode = {width = 0, height = 0,
    refresh = 0}, position = {x = 0, y = 0}, scale = 0, transform = WL_OUTPUT_TRANSFORM_NORMAL}
(gdb) p *matches[0]
$8 = {state = 0x7fffffffe370, wlr_head = 0x55555556ac10, link = {prev = 0x7fffffffe388,
    next = 0x55555556b930}, name = 0x55555555d780 "eDP-1",
  description = 0x55555556ac60 "Sharp Corporation 0x1484 0x00000000 (eDP-1)", phys_width = 290,
  phys_height = 170, modes = {prev = 0x55555556af90, next = 0x55555556acb0}, enabled = false,
  mode = 0x0, custom_mode = {width = 0, height = 0, refresh = 0}, x = 0, y = 0,
  transform = WL_OUTPUT_TRANSFORM_NORMAL, scale = 1}
(gdb) p *matches[1]
$9 = {state = 0x7fffffffe370, wlr_head = 0x555555562e60, link = {prev = 0x555555563d30,
    next = 0x7fffffffe388}, name = 0x555555562900 "DP-3",
  description = 0x55555555e450 "Dell Inc. DELL UP3216Q 6XF0D659376P (DP-3)", phys_width = 700,
  phys_height = 400, modes = {prev = 0x55555556a180, next = 0x555555563370}, enabled = true,
  mode = 0x55555556a170, custom_mode = {width = 0, height = 0, refresh = 0}, x = 0, y = 0,
  transform = WL_OUTPUT_TRANSFORM_NORMAL, scale = 2}
```